ipfs-retriever/lib/store.js

```diff
@@ -0,0 +1,392 @@
+ * @returns {{
+ *   serviceProviderId: string
+ *   serviceUrl: string
+ *   dataSetId: string
+ *   pieceId: string
+ *   ipfsRootCid: string
+ * }}
+ */
+function validateQueryResultsAndGetProvider(params) {
+  const { results, payerAddress, lookupKey } = params
+
+  httpAssert(
+    results && results.length > 0,
+    404,
+    `${lookupKey} does not exist or may not have been indexed yet.`,
+  )
+
+  const withServiceProvider = results.filter(
+    (row) => row && row.service_provider_id != null,
+  )
+  httpAssert(
+    withServiceProvider.length > 0,
+    404,
+    `${lookupKey} exists but has no associated service provider.`,
+  )
+
+  const withPaymentRail = withServiceProvider.filter(
+    (row) =>
+      row.payer_address && row.payer_address.toLowerCase() === payerAddress,
+  )
+  httpAssert(
+    withPaymentRail.length > 0,
+    402,
+    `There is no Filecoin Warm Storage Service deal for payer '${payerAddress}' and ${lookupKey}.`,
+  )
+
+  const withCDN = withPaymentRail.filter(
+    (row) => row.with_cdn && row.with_cdn === 1,
+  )
+  httpAssert(
+    withCDN.length > 0,
+    402,
+    `The Filecoin Warm Storage Service deal for payer '${payerAddress}' and ${lookupKey} has withCDN=false.`,
+  )
+
+  const withIpfsIndexing = withCDN.filter((row) => row.with_ipfs_indexing === 1)
+  httpAssert(
+    withIpfsIndexing.length > 0,
+    402,
+    `The Filecoin Warm Storage Service deal for payer '${payerAddress}' and ${lookupKey} has withIpfsIndexing=false.`,
+  )
+
+  const withPayerNotSanctioned = withIpfsIndexing.filter(
+    (row) => !row.is_sanctioned,
+  )
+  httpAssert(
+    withPayerNotSanctioned.length > 0,
+    403,
+    `Wallet '${payerAddress}' is sanctioned and cannot retrieve ${lookupKey}.`,
+  )
+
+  const withApprovedProvider = withPayerNotSanctioned.filter(
+    (row) => row.service_url,
+  )
+  httpAssert(
+    withApprovedProvider.length > 0,
+    404,
+    `No approved service provider found for payer '${payerAddress}' and ${lookupKey}.`,
+  )
+
+  const withIpfsRootCid = withApprovedProvider.filter(
+    (row) => row.ipfs_root_cid,
+  )
+  httpAssert(
+    withIpfsRootCid.length > 0,
+    404,
+    `${lookupKey} exists but has no associated IPFS Root CID.`,
+  )
+
+  const {
+    piece_id: pieceId,
+    data_set_id: dataSetId,
+    ipfs_root_cid: ipfsRootCid,
+    service_provider_id: serviceProviderId,
+    service_url: serviceUrl,
+  } = withApprovedProvider[0]
+
+  // We need this assertion to supress TypeScript error. The compiler is not able to infer that
+  // `withApprovedProvider.filter()` above returns only rows with `service_url` defined.
+  httpAssert(serviceUrl, 500, 'should never happen')
+
+  console.log(
+    `Validated data set ID '${dataSetId}', piece ID '${pieceId}', and service provider id '${serviceProviderId}' for ${lookupKey} and payer '${payerAddress}'. Service URL: ${serviceUrl}`,
+  )
+
+  return { serviceProviderId, serviceUrl, dataSetId, pieceId, ipfsRootCid }
```

Comment:
We will need to modify this to follow the changes made in https://github.com/filbeam/worker/pull/438